Empty folder in web ui and via 'mc ls', logs report unable to fetch object info

## Expected Behavior
Viewing a bucket/folder in the web UI or via 'mc ls' should show all the files that I know to be in that folder.

## Current Behavior
View two of the larger folders on my minio instance (not even that large, just a gig or so of data) shows no contents. The folders the minio instance is mounted onto show the files as expected. The logs report

```
2018-03-26T12:04:56.509471312Z app[web.1]: [2018-03-26T12:04:56.508591309Z] [ERROR] Unable to fetch object info for minecraft/backup-20180306.tgz (corrupted backend format, please join https://slack.minio.io for assistance)
2018-03-26T12:05:28.059937673Z app[web.1]:
2018-03-26T12:05:28.060002387Z app[web.1]: Trace: 1: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/fs-v1.go:983:cmd.(*FSObjects).ListObjects()
2018-03-26T12:05:28.060021074Z app[web.1]:        2: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/web-handlers.go:260:cmd.(*webAPIHandlers).ListObjects()
2018-03-26T12:05:28.060037927Z app[web.1]:        3: /opt/go/src/reflect/value.go:434:reflect.Value.call()
2018-03-26T12:05:28.060053727Z app[web.1]:        4: /opt/go/src/reflect/value.go:302:reflect.Value.Call()
2018-03-26T12:05:28.060069587Z app[web.1]:        5: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/gorilla/rpc/v2/server.go:137:v2.(*Server).ServeHTTP()
2018-03-26T12:05:28.060086259Z app[web.1]:        6: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP()
2018-03-26T12:05:28.060128907Z app[web.1]:        7: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:617:cmd.rateLimit.ServeHTTP()
2018-03-26T12:05:28.060145789Z app[web.1]:        8: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:575:cmd.pathValidityHandler.ServeHTTP()
2018-03-26T12:05:28.060161649Z app[web.1]:        9: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:514:cmd.httpStatsHandler.ServeHTTP()
2018-03-26T12:05:28.060177279Z app[web.1]:       10: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:65:cmd.requestSizeLimitHandler.ServeHTTP()
2018-03-26T12:05:28.060192818Z app[web.1]:       11: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:90:cmd.requestHeaderSizeLimitHandler.ServeHTTP()
2018-03-26T12:05:28.060208398Z app[web.1]:       12: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP()
2018-03-26T12:05:28.060224488Z app[web.1]:       13: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:231:cmd.redirectHandler.ServeHTTP()
2018-03-26T12:05:28.060240188Z app[web.1]:       14: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:289:cmd.minioReservedBucketHandler.ServeHTTP()
2018-03-26T12:05:28.060255637Z app[web.1]:       15: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:258:cmd.cacheControlHandler.ServeHTTP()
2018-03-26T12:05:28.060271136Z app[web.1]:       16: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:359:cmd.timeValidityHandler.ServeHTTP()
2018-03-26T12:05:28.060286505Z app[web.1]:       17: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1()
2018-03-26T12:05:28.060301854Z app[web.1]:       18: /opt/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP()
2018-03-26T12:05:28.060320601Z app[web.1]:       19: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:465:cmd.resourceHandler.ServeHTTP()
2018-03-26T12:05:28.060336421Z app[web.1]:       20: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/auth-handler.go:265:cmd.authHandler.ServeHTTP()
2018-03-26T12:05:28.060351920Z app[web.1]:       21: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/generic-handlers.go:133:cmd.reservedMetadataHandler.ServeHTTP()
2018-03-26T12:05:28.060367289Z app[web.1]:       22: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/http/server.go:111:http.(*Server).Start.func1()
2018-03-26T12:05:28.060388874Z app[web.1]:       23: /opt/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP()
2018-03-26T12:05:28.060405676Z app[web.1]:       24: /opt/go/src/net/http/server.go:2619:http.serverHandler.ServeHTTP()
2018-03-26T12:05:28.060421406Z app[web.1]:       25: /opt/go/src/net/http/server.go:1801:http.(*conn).serve()
```

## Possible Solution
I removed the .minio-sys/buckets/backups/minecraft folder and now my files show as expected. The folder I removed has not reappeared (as yet)

## Steps to Reproduce (for bugs)
* Have corrupt object files(?!)
* See no files in UI or via CLI
* Known corrupted files in the attached tgz are 20180305.tgz and 20180306.tgz
[minecraftbuckets.zip](https://github.com/minio/minio/files/1848208/minecraftbuckets.zip)

## Context
Ideally the system should be able to fix corrupted object files like this.

## Your Environment
* Version used (`minio version`): 2018-03-19T19:22:06Z
* Environment name and version (e.g. nginx 1.9.1): nginx/1.13.10
* Server type and version: Ubuntu 16.04
* Operating System and version (`uname -a`): Linux dokku 4.15.5-mainline-rev1 #1 SMP Thu Feb 22 16:22:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

--- a/cmd/fs-v1.go
+++ b/cmd/fs-v1.go
@@ -25,6 +25,7 @@ import (
 	"os"
 	"path"
 	"sort"
+	"strings"
 	"sync"
 	"time"
 
@@ -32,8 +33,12 @@ import (
 	"github.com/minio/minio/pkg/hash"
 	"github.com/minio/minio/pkg/lock"
 	"github.com/minio/minio/pkg/madmin"
+	"github.com/minio/minio/pkg/mimedb"
 	"github.com/minio/minio/pkg/policy"
 )
+
+// Default etag is used for pre-existing objects.
+var defaultEtag = "00000000000000000000000000000000-1"
 
 // FSObjects - Implements fs object layer.
 type FSObjects struct {
@@ -519,12 +524,48 @@ func (fs *FSObjects) getObject(ctx context.Context, bucket, object string, offse
 	return toObjectErr(err, bucket, object)
 }
 
+// Create a new fs.json file, if the existing one is corrupt. Should happen very rarely.
+func (fs *FSObjects) createFsJSON(object, fsMetaPath string) error {
+	fsMeta := newFSMetaV1()
+	fsMeta.Meta = make(map[string]string)
+	fsMeta.Meta["etag"] = GenETag()
+	if objectExt := path.Ext(object); objectExt != "" {
+		if content, ok := mimedb.DB[strings.ToLower(strings.TrimPrefix(objectExt, "."))]; ok {
+			fsMeta.Meta["content-type"] = content.ContentType
+		} else {
+			fsMeta.Meta["content-type"] = "application/octet-stream"
+		}
+	}
+	wlk, werr := fs.rwPool.Create(fsMetaPath)
+	if werr == nil {
+		_, err := fsMeta.WriteTo(wlk)
+		wlk.Close()
+		return err
+	}
+	return werr
+}
+
+// Used to return default etag values when a pre-existing object's meta data is queried.
+func (fs *FSObjects) defaultFsJSON(object string) fsMetaV1 {
+	fsMeta := newFSMetaV1()
+	fsMeta.Meta = make(map[string]string)
+	fsMeta.Meta["etag"] = defaultEtag
+	if objectExt := path.Ext(object); objectExt != "" {
+		if content, ok := mimedb.DB[strings.ToLower(strings.TrimPrefix(objectExt, "."))]; ok {
+			fsMeta.Meta["content-type"] = content.ContentType
+		} else {
+			fsMeta.Meta["content-type"] = "application/octet-stream"
+		}
+	}
+	return fsMeta
+}
+
 // getObjectInfo - wrapper for reading object metadata and constructs ObjectInfo.
 func (fs *FSObjects) getObjectInfo(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
 	fsMeta := fsMetaV1{}
 	fi, err := fsStatDir(ctx, pathJoin(fs.fsPath, bucket, object))
 	if err != nil && err != errFileAccessDenied {
-		return oi, toObjectErr(err, bucket, object)
+		return oi, err
 	}
 	if fi != nil {
 		// If file found and request was with object ending with "/", consider it
@@ -532,8 +573,7 @@ func (fs *FSObjects) getObjectInfo(ctx context.Context, bucket, object string) (
 		if hasSuffix(object, slashSeparator) {
 			return fsMeta.ToObjectInfo(bucket, object, fi), nil
 		}
-		logger.LogIf(ctx, errFileNotFound)
-		return oi, toObjectErr(errFileNotFound, bucket, object)
+		return oi, errFileNotFound
 	}
 
 	fsMetaPath := pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket, object, fs.metaJSONFile)
@@ -543,34 +583,35 @@ func (fs *FSObjects) getObjectInfo(ctx context.Context, bucket, object string) (
 	rlk, err := fs.rwPool.Open(fsMetaPath)
 	if err == nil {
 		// Read from fs metadata only if it exists.
-		defer fs.rwPool.Close(fsMetaPath)
-		if _, rerr := fsMeta.ReadFrom(ctx, rlk.LockedFile); rerr != nil {
-			// `fs.json` can be empty due to previously failed
-			// PutObject() transaction, if we arrive at such
-			// a situation we just ignore and continue.
-			if rerr != io.EOF {
-				return oi, toObjectErr(rerr, bucket, object)
-			}
+		_, rerr := fsMeta.ReadFrom(ctx, rlk.LockedFile)
+		fs.rwPool.Close(fsMetaPath)
+		if rerr != nil {
+			return oi, rerr
 		}
+	}
+
+	// Return a default etag and content-type based on the object's extension.
+	if err == errFileNotFound {
+		fsMeta = fs.defaultFsJSON(object)
 	}
 
 	// Ignore if `fs.json` is not available, this is true for pre-existing data.
 	if err != nil && err != errFileNotFound {
 		logger.LogIf(ctx, err)
-		return oi, toObjectErr(err, bucket, object)
+		return oi, err
 	}
 
 	// Stat the file to get file size.
 	fi, err = fsStatFile(ctx, pathJoin(fs.fsPath, bucket, object))
 	if err != nil {
-		return oi, toObjectErr(err, bucket, object)
+		return oi, err
 	}
 
 	return fsMeta.ToObjectInfo(bucket, object, fi), nil
 }
 
-// GetObjectInfo - reads object metadata and replies back ObjectInfo.
-func (fs *FSObjects) GetObjectInfo(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
+// getObjectInfoWithLock - reads object metadata and replies back ObjectInfo.
+func (fs *FSObjects) getObjectInfoWithLock(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
 	// Lock the object before reading.
 	objectLock := fs.nsMutex.NewNSLock(bucket, object)
 	if err := objectLock.GetRLock(globalObjectTimeout); err != nil {
@@ -587,6 +628,27 @@ func (fs *FSObjects) GetObjectInfo(ctx context.Context, bucket, object string) (
 	}
 
 	return fs.getObjectInfo(ctx, bucket, object)
+}
+
+// GetObjectInfo - reads object metadata and replies back ObjectInfo.
+func (fs *FSObjects) GetObjectInfo(ctx context.Context, bucket, object string) (oi ObjectInfo, e error) {
+	oi, err := fs.getObjectInfoWithLock(ctx, bucket, object)
+	if err == errCorruptedFormat || err == io.EOF {
+		objectLock := fs.nsMutex.NewNSLock(bucket, object)
+		if err = objectLock.GetLock(globalObjectTimeout); err != nil {
+			return oi, toObjectErr(err, bucket, object)
+		}
+
+		fsMetaPath := pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket, object, fs.metaJSONFile)
+		err = fs.createFsJSON(object, fsMetaPath)
+		objectLock.Unlock()
+		if err != nil {
+			return oi, toObjectErr(err, bucket, object)
+		}
+
+		oi, err = fs.getObjectInfoWithLock(ctx, bucket, object)
+	}
+	return oi, toObjectErr(err, bucket, object)
 }
 
 // This function does the following check, suppose

